Source/Common/Public/Future.h

```diff
@@ -0,0 +1,104 @@
+/*
+Copyright 2019 Chenxi Xu (@xsun2001)
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+
+#pragma once
+
+#include <atomic>
+#include <chrono>
+
+#include "Error.h"
+#include "SGEString.h"
+#include "Concurrent/Lock.h"
+
+namespace SpaceGameEngine
+{
+
+	struct ExecutionError
+	{
+		inline static const TChar sm_pContent[] = SGE_TSTR( "Execution Error" );
+
+		static bool Judge( const bool isFailed )
+		{
+			return isFailed;
+		}
+	};
+
+	struct TimeoutError
+	{
+		inline static const TChar sm_pContent[] = SGE_TSTR( "Timeout" );
+
+		static bool Judge( const bool isTimeout )
+		{
+			return isTimeout;
+		}
+	};
+
+	template<typename ResultType>
+	class Future
+	{
+		Mutex m_Mutex;
+		ResultType m_Result;
+		std::atomic_int8_t m_State;
```

Comment:
need we wrap atomic types?

Reply:
I just **use** atomic int. It is essential in concurrent environment.